cli/flox-watchdog/src/logger.rs

```diff
@@ -40,3 +48,245 @@ pub(crate) fn init_logger(file_path: &Option<PathBuf>) -> Result<(), anyhow::Err
         .init();
     Ok(())
 }
+
+/// Starts a background thread which emits a log entry at an interval. This is
+/// used as an indication of whether a watchdog's log file can be garbage
+/// collected. The thread will run until the watchdog exits.
+pub(crate) fn spawn_heartbeat_log() {
+    spawn(move || loop {
+        debug!("still watching, woof woof");
+        sleep(HEARTBEAT_INTERVAL);
+    });
+}
+
+/// Starts a background thread which garbage collects known log files. This is
+/// done on a best effort basis; errors are traced rather than being bubbled up
+/// and the thread will run until the watchdog exits.
+pub(crate) fn spawn_gc_logs(dir: impl AsRef<Path>) {
+    let dir = dir.as_ref().to_path_buf();
+    std::thread::spawn(move || {
+        gc_logs_watchdog(&dir, KEEP_WATCHDOG_DAYS)
+            .unwrap_or_else(|err| error!(%err, "failed to delete watchdog logs"));
+        gc_logs_services(&dir, KEEP_SERVICES_LAST)
+            .unwrap_or_else(|err| error!(%err, "failed to delete services logs"));
+    });
+}
+
+/// Garbage collects watchdog log files, keeping the last N days by modified
+/// time. This relies on the watchdog emitting a heartbeat log file from
+/// `log_heartbeat`.
+fn gc_logs_watchdog(dir: impl AsRef<Path>, keep_days: u64) -> Result<()> {
+    let mut files = glob_log_files(dir, "watchdog.*.log")?;
+    let threshold = duration_from_days(keep_days);
+    let now = SystemTime::now();
+
+    // Defaults to keeping if mtime is not supported by platform or filesystem.
+    files.retain(|file| file_older_than(file, now, threshold).unwrap_or(false));
+
+    for file in files {
+        try_delete_log(file);
+    }
+
+    Ok(())
+}
+
+/// Garbage collects services log files, keeping the last N files by filename.
+/// This relies on the log files having a timestamp in their filename.
+fn gc_logs_services(dir: impl AsRef<Path>, keep_last: usize) -> Result<()> {
+    let mut files = glob_log_files(dir, "services.*.log")?;
+    if files.len() <= keep_last {
+        return Ok(());
+    }
+
+    files.sort_unstable();
+    if files.len() > keep_last {
+        files.truncate(files.len() - keep_last);
+    }
```

Comment:
some nits / observations: feel free to ignore


1) I think the files come out sorted already from the FS but i guess for clairty we can sort them just in case?
2) we already returned iff `files.len() <= keep_last` above

Reply:
Removed the additional conditional. I'd like to keep the sort for safety/clarity though, since we're dependent on the order.